openfisca_core/simulations.py

```diff
@@ -106,13 +106,15 @@ def calculate(self, variable_name, period):
             period = periods.period(period)
 
         self.tracer.enter_calculation(variable_name, period)
+        self.tracer.record_start(time.time_ns() / (10**9))
```

Comment:
I'm in favour of moving `time.time_ns()` to the tracer but:
* we need to test how sub-calculations measure their durations (aka `test_tracers.py::test_calculation_time_with_depth`)
* time precision / os clocks differ (so, a simulation on bulk data might need some other time unit) 🤔

Reply:
`record_start` and `enter_calculation` have specific and different missions so I would keep them separated. 

> Besides, this is a first flame chart test based on timestamps but the chart values might take other performance information like the number of calls to the same variable which leads to a call to `enter_calculation` but no`record_start` (Even if, I know, it's bad to involve a possible future to a PR 😬).

Reply:
> `record_start` and `enter_calculation` have specific and different missions so I would keep them separated. 

Keeping them separated as 2 different functions in the `tracers` module makes total sense 👍 . 

But do they really have different missions from the simulation's point of view ? The fact that the simulation needs to tell the tracer to first enter a calculation, then to record the time feels to me like the simulation is telling the tracer how to do its job. And this makes me feel like it's an SRP violation.

I also find that the names read very similar at first sight:
- `tracer.record_start `: Record that something has started. Probably a calculation? 
- `tracer.enter_calculation`: Make the tracer enter a calculation. So tell the tracer a calculation has started, how is this different?

> I don't really understand how capturing more information would lead to call `enter_calculation` but not `record_start`, since we decided call `record_start` in all cases, even if we are just reading from the cache. 

This is not a hill I'm willing to die on, but I think we could have a single exposed method explicitly called `record_calculation_start` that would then call private methods to both `record_start_time` and `enter_calculation`.

Reply:
I added a commit containing [the 1st suggested change](https://github.com/openfisca/openfisca-core/pull/895#discussion_r315853136) on this branch, and [another commit](https://github.com/openfisca/openfisca-core/commit/60b3b3b338a21f6486d0f690d1422e7ced4b2c40) addressing my [2nd remark](https://github.com/openfisca/openfisca-core/pull/895#discussion_r315857967)
on a separate branch `trace-performance-suggestion` . 

Feel free to adopt or drop the second one (I think it simplifies the tracer interface and keep responsibilities more separated, but up to you).